crates/qt-gui/src/image_painter.rs

```diff
@@ -0,0 +1,170 @@
+            file_url: QUrl::default(),
+            running: false,
+        }
+    }
+}
+
+impl cxx_qt::Initialize for qobject::ImagePainter {
+    fn initialize(mut self: Pin<&mut Self>) {
+        self.as_mut().set_fill_color(&QColor::from_rgb(255, 0, 0));
+
+        self.as_mut()
+            .on_filter_changed(|this| {
+                this.load_file();
+            })
+            .release();
+
+        self.as_mut()
+            .on_file_url_changed(|this| {
+                this.load_file();
+            })
+            .release();
+    }
+}
+
+impl qobject::ImagePainter {
+    unsafe fn paint(self: Pin<&mut Self>, painter: *mut qobject::QPainter) {
+        // We need to convert the *mut QPainter to a Pin<&mut QPainter> so that we can reach the methods
+        if let Some(painter) = painter.as_mut() {
+            let pinned_painter = Pin::new_unchecked(painter);
+
+            // Now pinned painter can be used as normal
+            if let Some(image) = self.image.as_ref() {
+                let size = self.size();
+                pinned_painter
+                    .draw_image(&QRectF::new(0.0, 0.0, size.width(), size.height()), image);
+            }
+        }
+    }
+
+    fn do_load_file(path: &str, filter: &str) -> Result<QImage, Box<dyn Error>> {
+        let filter = filter.to_string().parse().unwrap();
+        let image = image::open(path)?.apply_filter(filter);
+
+        let mut bytes = QByteArrayCursor::new(QByteArray::default());
+        image.write_to(&mut bytes, image::ImageOutputFormat::Png)?;
+
+        QImage::from_data(bytes.as_ref(), "PNG").ok_or("Failed to convert to QImage!".into())
+    }
+
+    fn load_file(mut self: Pin<&mut Self>) {
+        if self.running {
+            return;
+        }
+
+        if let Some(local_file) = self
+            .file_url
+            .to_local_file()
+            .as_ref()
+            .map(ToString::to_string)
+        {
+            let file_name = self.file_url.file_name();
+            let filter = self.filter.to_string();
+
+            let qt_thread = self.as_mut().qt_thread();
+            self.as_mut().set_running(true);
+
+            std::thread::spawn(move || match Self::do_load_file(&local_file, &filter) {
+                Ok(image) => {
+                    qt_thread
+                        .queue(move |mut this| {
+                            this.as_mut().set_running(false);
+                            this.as_mut().set_status(file_name);
+                            this.as_mut().rust_mut().image = Some(image);
+                            this.as_mut().update();
+                        })
+                        .ok();
+                }
+                Err(err) => {
+                    let error = QString::from(&format!("Error: {}", err));
+                    qt_thread
+                        .queue(|mut this| {
+                            this.as_mut().set_running(false);
+                            this.set_status(error);
```

Comment:
do you need to set None on the image and call and update on the QQuickPaintedItem here ? Or are you leaving the old image there when an error occurs

Reply:
Ah, hm, good point.
Currently it's leaving the old image there.
But as we're clearing the status (including the old file name), we should probably also clear the image.

Reply:
Done